### DISCUSSION TITLE
docs/events

### DISCUSSION THREAD

**Discussion by giscus** [mikro-orm/mikro-orm#5783] [Comments]
# docs/events

There are two ways to hook to the lifecycle of an entity:

https://mikro-orm.io/docs/events

**Comment by crackedbeefcake** [comment]
hey,

is there a way to hook into a query?
i would like to attach a parameter to the `where` statement on every query (and possible all CRUD).

ilke this:
```
Mikro.init({
    ...
    hooks:{
      onQuery:(args)=> {
        args.where.runtime = runtimeId;
        return args;
      }
    }
})
```

**Reply by B4nan** [reply]
You can define a global filter:

https://mikro-orm.io/docs/filters

We also have `onQuery` hook but that is much more low level, it gets a raw SQL.

**Comment by NfNitLoop** [comment]
EventArgs has an EntityManager. But where in the lifecycle is the EventManager during an event handler?

For example, for AfterCreate or AfterUpdate, are we still within the same transaction that triggered the hook? Can we just modify/create/persist new objects within the hook and expect that they'll get stored within that transaction?

Or, do we have to explicitly call .flush() (possibly by multiple triggers) which gives a new database transaction and round trip?

**Reply by B4nan** [reply]
both events are executed inside the transaction. this very page you are commenting on contains all the info you need for adding more stuff to the current flush - the `onFlush` event in particular

**Reply by NfNitLoop** [reply]
Thanks for the fast response! 

> both events are executed inside the transaction

Does that go for the implicit transaction that an `em.flush()` usually operates within? That would be nice!

I did see onFlush too -- I was just wondering if the other event handlers would happen in the same flush/transaction, since I didn't see it explicitly called out. 

I'll likely use onFlush since I can handle CREATE/UPDATE(/UPSERT?) in one method. One part of the docs say:

> You can **either** register the subscribers manually in the ORM configuration 

I think the **or** for that either/or statement got lost in docs. Is there another way to register the subscriber? (It would be more convenient than manually adding all of my subscribers into the global config.)

**Reply by B4nan** [reply]
> Does that go for the implicit transaction that an em.flush() usually operates within? That would be nice!

yes, they are triggered from the flush, everything is in the same transaction context

> I did see onFlush too -- I was just wondering if the other event handlers would happen in the same flush/transaction, since I didn't see it explicitly called out.

they will, but they are executed after all changesets are computed and grouped - so you cant just add stuff via `em.persist`, you need to compute the changeset manually as described in the `onFlush` docs

> I think the or for that either/or statement got lost in docs. Is there another way to register the subscriber?

you can either register them globally in the ORM config, or manually via `em.getEventManager().registerSubscriber()` - that way you can register a subscriber on a particular fork instead of having it globally available on all EM forks. the docs are indeed missing the second approach. its mentioned [here](https://mikro-orm.io/docs/usage-with-nestjs#using-eventsubscriber) btw, its especially handy for nestjs where you can create the subscriber via nest di, and have some additional dependencies injected, and self-register it via the constructor.
